### PR TITLE
PPM Screen Changes

### PR DESCRIPTION
**Pull request by KaushikMalapati**
## Description

Changing responsivity io to be input only so it displays without a connection error
Renaming manual pvs to be less similar to auto background pvs
Adding microjoule mirror of calibrated_mj
Added manual collection in progress indicator
Replaced references to passive mode with auto mode

## Motivation and Context

https://jira.slac.stanford.edu/browse/ECS-875

## How Has This Been Tested?

Interactively

## Where Has This Been Documented?

upcoming-release notes

image

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [ ] Test suite passes on GitHub Actions
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

File: pcdsdevices/pim.py [hunk 1]
```diff
@@ -372,32 +372,43 @@ class PPMPowerMeter(BaseInterface, Device):
...
     is used to calculate the actual energy in the following way:
 
-    `calibrated_mj` = 1000 * (Signal - Background) / (Responsivity * Beam_Rate)
-    """
+    `calibrated_mj` = (Signal - Background) / (Responsivity * Beam_Rate)
+    """
+    def __init__(self, prefix, *, name, **kwargs):
+        import sympy.physics.units as units
+        from sympy.physics.units.prefixes import micro, milli
+        from sympy.physics.units.quantities import Quantity
```
Comment (tangkong) [inline pcdsdevices/pim.py:380]: We should avoid importing modules inside of class instance methods, unless there's a very specific reason not to.  This is a base class that gets inherited in many later devices, and we'll be re-importing these modules each time one of these classes is instantiated

File: pcdsdevices/pim.py [hunk 2]
```diff
@@ -372,32 +372,43 @@ class PPMPowerMeter(BaseInterface, Device):
...
+    responsivity = Cpt(PytmcSignal, ':RES', io='i', kind='normal', doc='Responsivity in  V/W, unique for every power meter.')
 
     background_voltage = Cpt(PytmcSignal, ':BACK:VOLT', io='io', kind='normal', doc='Background voltage value used to calculate pulse energy.')
 
     auto_background_reset = Cpt(PytmcSignal, ':BACK:RESET', io='io', kind='normal', doc='Set to reset auto background voltage collection.')
     set_metadata(auto_background_reset, dict(variety='command-proc', value=1))
 
-    background_mode = Cpt(PytmcSignal, ':BACK:MODE', io='io', kind='normal', doc='Can be manual or passive. In manual mode, you can collect for a specified number of seconds. In passive mode, a buffer of automatically collected background voltages will be used to calculate the background voltage.')
+    background_mode = Cpt(PytmcSignal, ':BACK:MODE', io='io', kind='normal', doc='Can be manual or auto In manual mode, you can collect for a specified number of seconds. In auto mode, a buffer of automatically collected background voltages will be used to calculate the background voltage.')
```
Comment (tangkong) [inline pcdsdevices/pim.py:398]: Can we break up some of these lines, they are very long

File: pcdsdevices/pim.py [hunk 3]
```diff
@@ -372,32 +372,43 @@ class PPMPowerMeter(BaseInterface, Device):
...
+    """
+    def __init__(self, prefix, *, name, **kwargs):
+        import sympy.physics.units as units
+        from sympy.physics.units.prefixes import micro, milli
+        from sympy.physics.units.quantities import Quantity
+        millijoule = mJ = Quantity("millijoule", abbrev='mJ')
+        millijoule.set_global_relative_scale_factor(milli, units.joule)
+        microjoule = uJ = Quantity("microjoule", abbrev='uJ')
+        microjoule.set_global_relative_scale_factor(micro, units.joule)
+        setattr(units, 'mJ', mJ)
```
Comment (tangkong) [inline pcdsdevices/pim.py:385]: So what we're doing here is manually adding a custom unit to the `sympy.physics.units` module?  This is almost certainly now how sympy is meant to be used, and import order can make this inconsistent.

The definition of these should probably go into `pcdsdevices.utils`, with maybe a modification of how we acquire units (First look for native sympy units, then look for our custom defined units)

File: pcdsdevices/pim.py (hunk 1 above)
Comment (KaushikMalapati) [reply pcdsdevices/pim.py]: I removed all of this from the class

**Comment by KaushikMalapati** [comment]
@tangkong, do you know if there's a way to remove the lineedit by calibrated_uj and the two circle indicators by the command buttons for auto_background_reset and manual_collect? calibrated_uj already has write_access=False and the two command indicators are not useful because they are reset to 0 on the cycle when they are set to one, so the screen never gets updates the circles. These would be purely cosmetic changes.

**Comment by tangkong** [comment]
Not to my knowledge, as far as I know the autogenerated screens don't let you pick and choose which parts of the widget get displayed.  You could make a custom template for your screens, but that does seem like overkill

**Review by tangkong — COMMENTED** [review]
I like how this turned out, I think the implementation of our custom units turned out quite clean.

Do you feel up to writing a quick unit test for the new `convert_unit` function?  You could make a new file (`test_units.py`) and just write a test that converts between mJ, uJ, and J.  We don't have a test for this yet and it'll be good to make sure this function stays functional

File: pcdsdevices/pim.py [hunk 4]
```diff
@@ -372,24 +372,38 @@ class PPMPowerMeter(BaseInterface, Device):
     calibrated for each power meter in units of volts per watt
     is used to calculate the actual energy in the following way:
 
-    `calibrated_mj` = 1000 * (Signal - Background) / (Responsivity * Beam_Rate)
+    `calibrated_mj` = (Signal - Background) / (Responsivity * Beam_Rate)
     """
+    def __init__(self, prefix, *, name, **kwargs):
+        super().__init__(prefix, name=name, **kwargs)
```
Comment (tangkong) [inline pcdsdevices/pim.py:378]: Just realizing this is here, we don't really need this init do we?

Comment (KaushikMalapati) [reply pcdsdevices/pim.py]: No

**Comment by KaushikMalapati** [comment]
Changed the unit test to use the long unit names and added a shorter one for sympy units

**Review by tangkong — APPROVED** [review]

**Comment by KaushikMalapati** [comment]
@tangkong Can I git pull origin master in /cds/group/pcds/pyps/apps/dev/pcdsdevices to update pcdsdevices for dev_conda?

**Comment by KaushikMalapati** [comment]
> > a way to remove the lineedit by calibrated_uj and the two circle indicators by the command buttons for auto_background_reset and manual_collect? calibrated_uj already has write_access=False and the two command indicators are not useful because they are reset to 0 on the cycle when they are set to one, so the screen never gets updates the circles. These would be purely cosmetic changes.
> 
> Would you find it useful for us to add a way to do this using https://github.com/pcdshub/pcdsdevices/blob/master/pcdsdevices/variety.py and https://github.com/pcdshub/typhos/blob/master/typhos/variety.py ?
> 
> Hypothetically we can come up with some variety tag as a message for typhos that indicates that we don't want circle indicators on our commands etc.

I don't think it's worth doing since it'd be pretty specific. I could always make a custom ui for it but I think leaving the indicator there is fine.

**Comment by ZLLentz** [comment]
Ok, no worries. I figured this might be a more common ask.

**Pull request merged**